app-backend/batch/src/main/scala/com/azavea/rf/batch/ast/package.scala

```diff
@@ -34,6 +34,70 @@ package object ast {
     TileLayerRDD(f(tlr0, tlr1), tlr0.metadata.combine(tlr1.metadata))
   }
 
+  /** Perform a binary reduction on the arguments of some Operation node.
+    * This is complicated by [[Constant]], which renders certain AST
+    * combinations meaningless when dealing with RDDs and GeoTrellis layers. For
+    * instance, how should one evaluate the AST:
+    * {{{
+    * Masking <- Constant
+    * }}}
+    * While this may have meaning on the Tile server, it is meaningless here,
+    * since there is no way to devise a key-space nor metadata for the RDD we'd
+    * have to pull from thin air.
+    *
+    * This function _looks_ like it can throw an error, but it never should.
+    * ASTs that contain Constant nodes in illegal positions will be rejected by
+    * the validator before this code here can ever run.
+    */
+  private def reduce(
```

Comment:
This is great. I wonder if you have some ideas about how we might be able to genericize structural checks like this in MAML